services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusyhteystietoEntity.kt

```diff
@@ -60,8 +59,18 @@ class HakemusyhteystietoEntity(
                         puhelin = yhteyshenkilo.hankekayttaja.puhelin,
                         tilaaja = yhteyshenkilo.tilaaja,
                     )
-                }
-        )
+                })
+
+    fun copyToHakemus(hakemus: HakemusEntity) =
+        HakemusyhteystietoEntity(
+            tyyppi = tyyppi,
+            rooli = rooli,
+            nimi = nimi,
+            sahkoposti = sahkoposti,
+            puhelinnumero = puhelinnumero,
+            ytunnus = ytunnus,
+            application = hakemus,
+            yhteyshenkilot = yhteyshenkilot.map { it.copyToYhteystieto(this) }.toMutableList())
```

Comment:
`this` is the pre-existing yhteystieto with the old hakemus, not the one being created here.

Reply:
Yep, this is a problem. The new yhteyshenkilöt are linked to the wrong yhteystieto.

Before send:
```
haitaton=# select * from hakemusyhteystieto where application_id=20;
                  id                  | application_id | tyyppi  |      rooli      |         nimi         |       sahkoposti        | puhelinnumero | y_tunnus
--------------------------------------+----------------+---------+-----------------+----------------------+-------------------------+---------------+-----------
 3b722902-f515-4864-85d2-2f5208c4c1a8 |             20 | COMPANY | HAKIJA          | Hakija Oy            | hakija@yritys.test      | 0411234567    | 1423830-0
 5103d9c6-c4f1-4262-acc3-c81d62797725 |             20 | COMPANY | TYON_SUORITTAJA | Suorittaja Oy        | suorittaja@yritys.test  | 0411234567    | 1423830-0
 e939add4-d94d-4dba-a15d-b357c2be6366 |             20 | COMPANY | RAKENNUTTAJA    | Rakennuttajafirma Oy | info@rakennuttajat.test | 0407654321    |
(3 rows)

haitaton=# select * from hakemusyhteyshenkilo where hakemusyhteystieto_id in (select id from hakemusyhteystieto where application_id = 20);
                  id                  |        hakemusyhteystieto_id         |           hankekayttaja_id           | tilaaja
--------------------------------------+--------------------------------------+--------------------------------------+---------
 785b228d-8901-41d3-b65c-aa5c71eb9ca3 | 3b722902-f515-4864-85d2-2f5208c4c1a8 | 5e18982f-cecb-4017-8a17-773df6d58bbe | f
 1af5eb5b-75c2-4262-a2d8-723685666d86 | 5103d9c6-c4f1-4262-acc3-c81d62797725 | d170f845-28a3-44c3-b177-f8bd43fadfb3 | f
 165c5e59-16b1-44e1-9cff-8f69c5799213 | e939add4-d94d-4dba-a15d-b357c2be6366 | 8e436605-7c3a-4f9b-bffc-bdbfd849760e | f
(3 rows)
```

After send:
```
haitaton=# select * from hakemusyhteystieto where application_id=20;
                  id                  | application_id | tyyppi  |      rooli      |         nimi         |       sahkoposti        | puhelinnumero | y_tunnus
--------------------------------------+----------------+---------+-----------------+----------------------+-------------------------+---------------+-----------
 3b722902-f515-4864-85d2-2f5208c4c1a8 |             20 | COMPANY | HAKIJA          | Hakija Oy            | hakija@yritys.test      | 0411234567    | 1423830-0
 5103d9c6-c4f1-4262-acc3-c81d62797725 |             20 | COMPANY | TYON_SUORITTAJA | Suorittaja Oy        | suorittaja@yritys.test  | 0411234567    | 1423830-0
 e939add4-d94d-4dba-a15d-b357c2be6366 |             20 | COMPANY | RAKENNUTTAJA    | Rakennuttajafirma Oy | info@rakennuttajat.test | 0407654321    |
(3 rows)

haitaton=# select * from hakemusyhteyshenkilo where hakemusyhteystieto_id in (select id from hakemusyhteystieto where application_id = 20);
                  id                  |        hakemusyhteystieto_id         |           hankekayttaja_id           | tilaaja
--------------------------------------+--------------------------------------+--------------------------------------+---------
 785b228d-8901-41d3-b65c-aa5c71eb9ca3 | 3b722902-f515-4864-85d2-2f5208c4c1a8 | 5e18982f-cecb-4017-8a17-773df6d58bbe | f
 165c5e59-16b1-44e1-9cff-8f69c5799213 | e939add4-d94d-4dba-a15d-b357c2be6366 | 8e436605-7c3a-4f9b-bffc-bdbfd849760e | f
 130e662b-4eef-4a6e-9e9d-9088682aa86e | 3b722902-f515-4864-85d2-2f5208c4c1a8 | 5e18982f-cecb-4017-8a17-773df6d58bbe | f
 b5112f19-e503-418a-a212-db44a904f41b | 5103d9c6-c4f1-4262-acc3-c81d62797725 | d170f845-28a3-44c3-b177-f8bd43fadfb3 | t
 47821479-0f4c-4e8c-ad34-4a4c7bad819f | e939add4-d94d-4dba-a15d-b357c2be6366 | 8e436605-7c3a-4f9b-bffc-bdbfd849760e | f
 1af5eb5b-75c2-4262-a2d8-723685666d86 | 5103d9c6-c4f1-4262-acc3-c81d62797725 | d170f845-28a3-44c3-b177-f8bd43fadfb3 | t
(6 rows)

haitaton=# select * from hakemusyhteystieto where application_id=21;
                  id                  | application_id | tyyppi  |      rooli      |         nimi         |       sahkoposti        | puhelinnumero | y_tunnus
--------------------------------------+----------------+---------+-----------------+----------------------+-------------------------+---------------+-----------
 e4dd5d53-2e73-4741-bcab-10422c2df9f1 |             21 | COMPANY | HAKIJA          | Hakija Oy            | hakija@yritys.test      | 0411234567    | 1423830-0
 5ca45379-93e1-479d-9263-ef755804f193 |             21 | COMPANY | TYON_SUORITTAJA | Suorittaja Oy        | suorittaja@yritys.test  | 0411234567    | 1423830-0
 60bbb1fe-7de4-45a3-9a2c-2f32cb629bd5 |             21 | COMPANY | RAKENNUTTAJA    | Rakennuttajafirma Oy | info@rakennuttajat.test | 0407654321    |
(3 rows)

haitaton=# select * from hakemusyhteyshenkilo where hakemusyhteystieto_id in (select id from hakemusyhteystieto where application_id = 21);
 id | hakemusyhteystieto_id | hankekayttaja_id | tilaaja
----+-----------------------+------------------+---------
(0 rows)
```

Reply:
Would this be a suitable fix to this:

```
    fun copyToHakemus(hakemus: HakemusEntity) =
        HakemusyhteystietoEntity(
            tyyppi = tyyppi,
            rooli = rooli,
            nimi = nimi,
            sahkoposti = sahkoposti,
            puhelinnumero = puhelinnumero,
            ytunnus = ytunnus,
            application = hakemus
        ).also { newEntity ->  
            newEntity.yhteyshenkilot.addAll(yhteyshenkilot.map { it.copyToYhteystieto(newEntity) })
        }
```

Reply:
I tested my proposed solutions above and it seems to work. Thanks a lot for noticing this!